### ISSUE TITLE
[FEATURE_REQUEST] text-generation-webui now has support for jinja chat templates

### ISSUE THREAD

**Issue by HiroseKoichi** [SillyTavern/SillyTavern#1538]
### Have you searched for similar requests?

Yes

### Is your feature request related to a problem? If so, please describe.

More complex prompt templates, such as chat-ML and LLama-2-chat, are unable to be properly used due to broken formatting, which leads to suboptimal output.

### Describe the solution you'd like

The [chat template](https://huggingface.co/docs/transformers/main/en/chat_templating) can be extracted from the model's metadata and used to ensure that the input is in the proper format. This would fix all current issues related to broken templates, and any new format would automatically be supported.

[text-generation-webui](https://github.com/oobabooga/text-generation-webui) has recently added support for jinja chat templates: https://github.com/oobabooga/text-generation-webui/pull/4874
I don't know if changes need to be made to text-gen-webui's API for SillyTavern to use it or if this is something that needs to be implemented manually, but I just wanted to bring it to your attention since chat template issues are becoming more frequent as more models are being released.

### Describe alternatives you've considered

You could add more options to the existing solution so that all templates can properly be handled, but this would still require new templates to be added every time a new format is created, and unlike with jinja templates, the user would still have to know what format the model was trained on and select the appropriate one.

### Priority

Medium (Would be very useful)

### Is this something you would be keen to implement?

No

**Comment by Cohee1207** [comment]
This can be done only if ooba source is duplicated for chat completion sources. Text completion loses the context about messages when building prompts.

**Comment by Dampfinchen** [comment]
You are quite right, both ChatML and Llama 2 Chat are not being followed correctly, which is a shame as those formats gain an increasing range of adoption, especially since Mixtral is using it too. People using SillyTavern won't get the full performance Mixtral and ChatML models have to offer or worse, are much more likely to impersonate or be incoherent. 

In my opinion, implementing these two formats correctly should receive top prioritiy above all, for obvious reasons. Extensions and other nice to have stuff can wait.

**Comment by Cohee1207** [comment]
@Dampfinchen 
These formats are not built around our use cases and are very restrictive.

- Free-form example messages in variable formats
- Chat messages without alternation of roles
- Free choice of who has the first turn
- Prompt injections at any depth and with any content

1. Don't adopt such formats.
2. Vote against them whenever possible.
3. Bring the attention of those who train models.

**Comment by Cohee1207** [comment]
@Dampfinchen So far all of the cases you've brought to me were due to using features that are not expected by the overly restrictive instruct formatting. If you want to use such a format - align your use accordingly.
* Remove all example dialogues, it's not a thing for these formats.
* Reject all injections (vectors, summaries, world info, etc), again not something it can accommodate.
* Start with the user message. `/send at=0 [New chat]` should shove a user message first.
* Don't use group chats or anything that produces non-alternating roles.

The only thing I can't offer an alternative for is having messages fall off from context in pairs.

**Comment by HiroseKoichi** [comment]
I personally don't like llama-2-chat at all, but chatML, on the other hand, is my favorite chat template. I don't think that it's restrictive; it just does things differently with the added benefit of clearly defining where the input and output are from.

> Remove all example dialogues, it's not a thing for these formats.

I've never had an issue using example messages in any format, so I'm a bit confused as to how you came to this conclusion.

> Reject all injections (vectors, summaries, world info, etc), again not something it can accommodate.

For chatML and likely others that have a clearly defined system prompt, you can just send these as a system prompt, and they work perfectly fine. For example, all I did with the author's notes was add this, then use it as normal:
```
<|im_end|>
<|im_start|>system
```

> Free choice of who has the first turn
>> Start with the user message. /send at=0 [New chat] should shove a user message first.

I've never had an issue with having the AI's message be first in any format, so once again, I'm confused as to how you came to this conclusion. Every single format has the user message come first, so I don't know why you're singling out these ones in particular.

> Chat messages without alternation of roles
>> Don't use group chats or anything that produces non-alternating roles.

Every single format does this, and no model has been trained otherwise. No matter what format I've used, having the AI generate two messages in a row gives very poor output because the model has output the EOS token signaling that it's confident it reached a complete answer, so making it generate a new response without additional input makes it do random stuff because it has nothing else to go off of. Alternatively, you can ban the EOS token and press Continue to make it generate more, but even then, it still goes off on a tangent.

The reason why group chats work at all despite not having an alternation of roles is because now the model actually does have additional information to go off of since it's supposed to respond as an entirely different character. I've done a few group chats using chatML, and it had no problem handling it.

---

On a side note, I've discovered that, when using chatML, adding ({{user}}) and ({{char}}) right next to the chatML roles is a much better way of showing who is what character rather than prefixing the user and AI messages with {{user}}: and {{char}}:

For user: 
```
<|im_start|>user ({{user}})
(user message)<|im_end|>
```
 instead of
```
<|im_start|>user
{{user}}: (user message)<|im_end|>
```

For AI:
```
<|im_start|>assistant ({{char}})
(char message)<|im_end|>
```
 instead of
```
<|im_start|>assistant
{{char}}: (char message)<|im_end|>
```
Unfortunately though, this doesn't appear to work properly in group chats since the macro {{char}} makes it so previous ones will be replaced with the current speaking bots name, and prefixing {{user}}: and {{char}}: is always enabled for group chats. Either way, groups chats have worked perfectly fine when using chatML.

**Comment by WolframRavenwolf** [comment]
Hi! Thought I'd chime in since I made those two presets. So yeah, they aren't perfect, just the best I could make with the options I had.

Llama 2 Chat is an abomination of a prompt template, making it extremely hard to implement properly, and I'd love to see that format die out. It's just hard to implement and when done exactly as intended, requires constant rewriting of the beginning of the context (putting the system message in the first user message is such terrible design). Plus, no way for the AI to go first.

But I've worked around it by making the system prompt the first message, so that doesn't change, and lets the AI go first as it will follow up on that with its greeting. Examples still suck, but at least it's usable in most situations.

With Mixtral's release, I've been using the Llama 2 Chat format unchanged a lot (except for enabling "Include Names" which is a good idea with most models), and it does work with Mixtral. Even though Mixtral has no official system prompt, it works very well, as it's a really smart model. Even group chat was no problem. (Side note: Mixtral is so smart that it can do group chats with a single char, and was the first model I've ever seen that did that on its own, bringing in a second character from the original persona's background, creating a perfect simulation of a group chat!)

ChatML is a much better format, and works pretty universally even with models not tuned on it, just like Alpaca or Vicuna. It's a little more complex, but has important advantages, like roles, distinct system messages, and special tokens that won't ever appear in normal text. Alpaca uses common markdown headers and Vicuna is just chat log format that's also very common. Those formats break easily when you do more than just chat with a model, e. g. when ingesting markdown or log files, whereas ChatML is still distinct and won't get the model confused. As we evolve our use cases and do more with local LLMs, we'll have to use a format that works in such advanced settings, and ChatML looks to be the best option that currently exists.

@HiroseKoichi Interesting findings with your placement of the char/user placeholders. I've also been wondering if we could just replace the "user/assistant" string after the BOS token with the char/user's name.

Looks like we need placeholders in previous messages to be literally replaced by their contents when they've been used for the first time (so they don't change with the "current" personality), or better yet, replacing them with character-specific placeholders (more effort to implement, but will handle renaming chars).

**Comment by Dampfinchen** [comment]
@HiroseKoichi That was private discourse I've had with Cohee, so it's natural you're confused. Let me explain.

Example dialogues with the Llama 2 Chat format are tricky in ST because they immediately add an [INST] after the character card, which should end with [/INST] instead. That is fixable by writing a [/INST] placeholder at the end of the character card though or just writing the example dialogue in plain text in the chara card. Same story with vector storage, but that is also fixable using a similar trick. 

However, what is not fixable is the major bug that appears right at the start of the chat messages in context. You see, according to Llama 2 chat, only the user response is surrounded by the input and output sequence ( [INST] ) and ( [/INST] ) This clashes with all other formats were both the user and bot are treated in the same way. This means in practice you have the input sequence [INST] appear at the beginning of the chat instead of [/INST] after the context is reached and the user response is first, which is occuring randomly. So the bot naturally gets confused. That happens regardless if you use example dialogues and vector storage or not. 

Aside from that, the ST formatting also introduces new lines after the character card, which the format, again, does not expect. 

So to put it simply, this is the proper Llama 2 chat template.

```
[INST] <<SYS>>
Write character's next reply.
<</SYS>>

Character card [/INST] {response 0} </s><s>[INST] {prompt 1} [/INST] {response 1} </s><s>...[INST] {prompt N} [/INST]
```

With Sillytavern formatting and its major bug happening it looks like this (after replacing the new line separator which is configured by ST as default for some reason with the correct one)

```
[INST] <<SYS>>
Write character's next reply.
<</SYS>>

Character card
</s><s>[INST] {prompt} [/INST] {response} </s><s>[INST] {prompt} [/INST] etc.
```
I imagine it's much easier to implement the ChatML format correctly than the Llama 2 Chat template. 

Very smart models like Mixtral should still work nicely even with the formatting errors. However, the performance will still be compromised due to this. According to Mistral, the prompt format has to be strictly followed for best performance.

**Comment by WolframRavenwolf** [comment]
@Dampfinchen The output you gave as an example of what SillyTavern produces isn't correct, though, at least not with the original preset. Did you modify yours in other ways than just replacing the Separator newline with a space? Here's what it looks for me with the original preset (last modified 2 months ago):

```
[INST] <<SYS>>
{{system}}
<</SYS>>

{{description}}
Scenario: {{scenario}}
***
[/INST] AI's greeting message
[INST] {prompt} [/INST] {response}
```

Ignoring the special tokens `</s>` and `<s>`, by replacing the Separator with a space, you get very close to the original Llama 2 Chat format (as long as you don't use Examples which add an unopened [/INST] after each example message - damn, I hate that format!). Unfortunately I can't remember why I made the Separator a linebreak instead of a space - I'm sure I had a reason for that, but right now, can't remember. 🤷

**Update:** Oops, my mistake, I didn't create that preset, I just fixed (some of it) and renamed it - the linebreak was there from the beginning when toby3d created the original preset. So I just didn't change that when I fixed some other issues.

**Comment by Dampfinchen** [comment]
> @Dampfinchen The output you gave as an example of what SillyTavern produces isn't correct, though, at least not with the original preset. Did you modify yours in other ways than just replacing the Separator newline with a space? Here's what it looks for me with the original preset (last modified 2 months ago):
> 
> ```
> [INST] <<SYS>>
> {{system}}
> <</SYS>>
> 
> {{description}}
> Scenario: {{scenario}}
> ***
> [/INST] AI's greeting message
> [INST] {prompt} [/INST] {response}
> ```
> 
> Ignoring the special tokens `</s>` and `<s>`, by replacing the Separator with a space, you get very close to the original Llama 2 Chat format (as long as you don't use Examples which add an unopened [/INST] after each example message - damn, I hate that format!). Unfortunately I can't remember why I made the Separator a linebreak instead of a space - I'm sure I had a reason for that, but right now, can't remember. 🤷

I did modify the template a ton but I can assure you, the bug also happens using the default template. I've tested that of course.

The thing is, it replaces the output sequence for the input sequence randomly and I've found out it only does so after reaching the maximum context (in my case 4096 tokens)

Could you please test it again on a long chat? You properly have to chat with the model multiple times for the bug to appear. 

My chat is pretty long (around 1300 messages) so perhaps that plays a role as well?

**Comment by WolframRavenwolf** [comment]
@Dampfinchen I just tested it with the orignal preset freshly downloaded, and checked the repo, Staging and the current release versions are identical. So that bug doesn't happen at first.

Oh, wait, you mean it happens only after context fills up? Yeah, that explains it, because the stupid Llama 2 Chat format can't have bot messages stand alone - the format has them always attached to the user message, and when the context fills up and older messages disappear, SillyTavern doesn't remove the message pair, just one message - breaking the format like you've seen. It appears random because it depends on when and how many messages scroll out of view, which is affected by message length and other settings, so it looks like it happens randomly.

That's why I always call it a "complicated" (and not just complex) prompt format - to implement it properly, you need a lot of extra prompt manipulation logic just for this one format. I wonder how other frontends deal with that? They'd have to work around all those issues, too.

**Comment by HiroseKoichi** [comment]
> @HiroseKoichi That was private discourse I've had with Cohee, so it's natural you're confused. Let me explain.

Ah, that makes sense.

> However, what is not fixable is the major bug that appears right at the start of the chat messages in context. You see, according to Llama 2 chat, only the user response is surrounded by the input and output sequence ( [INST] ) and ( [/INST] ) This clashes with all other formats were both the user and bot are treated in the same way. This means in practice you have the input sequence [INST] appear at the beginning of the chat instead of [/INST] after the context is reached and the user response is first, which is occuring randomly. So the bot naturally gets confused. That happens regardless if you use example dialogues and vector storage or not.

That sounds horrible to work with. I'm glad I haven't used it yet.

---

> @HiroseKoichi Interesting findings with your placement of the char/user placeholders. I've also been wondering if we could just replace the "user/assistant" string after the BOS token with the char/user's name.

This was one of the first things I tried to make it better at remembering who is playing what character, and while it was an improvement over prefixing {{user}}: and {{char}}:, the output felt a little off to me, so I decided that having the default user/assistant roles was necessary. With the new changes I've made, however, it hasn't mixed up who is playing what character or spoken for me. I could just need more time with it, but for now it looks really promising.

**Comment by WolframRavenwolf** [comment]
> This was one of the first things I tried to make it better at remembering who is playing what character, and while it was an improvement over prefixing {{user}}: and {{char}}:, the output felt a little off to me, so I decided that having the default user/assistant roles was necessary. With the new changes I've made, however, it hasn't mixed up who is playing what character or spoken for me. I could just need more time with it, but for now it looks really promising.

Ah, cool, saves me testing that. And reminds me of how the simple-proxy-for-tavern preset adds the char/user names. So definitely worth to experiment with.

**Comment by HiroseKoichi** [comment]
> This was one of the first things I tried to make it better at remembering who is playing what character, and while it was an improvement over prefixing {{user}}: and {{char}}:, the output felt a little off to me, so I decided that having the default user/assistant roles was necessary. With the new changes I've made, however, it hasn't mixed up who is playing what character or spoken for me. I could just need more time with it, but for now it looks really promising.

Looking back, I realize that I was terrible at prompting a few months ago when I tested this. I was using a character card that I had created myself, and looking at it now, it's utter garbage. I'm going to retest this with a few good ones to see whether I was right or whether I'm just stupid.

**Comment by HiroseKoichi** [comment]
>> This was one of the first things I tried to make it better at remembering who is playing what character, and while it was an improvement over prefixing {{user}}: and {{char}}:, the output felt a little off to me, so I decided that having the default user/assistant roles was necessary. With the new changes I've made, however, it hasn't mixed up who is playing what character or spoken for me. I could just need more time with it, but for now it looks really promising.
>>
> Looking back, I realize that I was terrible at prompting a few months ago when I tested this. I was using a character card that I had created myself, and looking at it now, it's utter garbage. I'm going to retest this with a few good ones to see whether I was right or whether I'm just stupid.



From my new tests, it looks any spaces in the role area is what was causing issues, but it can be fixed by adding parenthesis. For example, using `<|im_start|>({{user}})` and `<|im_start|>({{char}})` seems to provide the best results. If both your name and the characters name are one word, then the parenthesis is not needed, but if it's longer, such as a scenario name, then the parenthesis is needed.

I haven't tested too much with it, but it looks like this is giving the best results so far.

**Comment by HiroseKoichi** [comment]
Since it's unlikely that this will be implemented, I'm going to close this and reopen #1261.
